File: cpp/排序/堆排序.cpp
//堆排序 
//时间复杂度 O(N*logN)
 
#include <iostream>
using namespace std;
#define MAXSIZE 1000
int arr[MAXSIZE];
int n;

void swap(int l,int r)
{
	int temp = arr[l];
	arr[l] = arr[r];
	arr[r] = temp;
}

//i位置的数，向上调整大根堆
void heapInsert(int i)
{
	while(arr[i] > arr[(i-1)/2])
	{
		swap(i,(i-1)/2);
		i = (i-1)/2;
	}
} 

//i位置的数，向下调整大根堆
//当前堆的大小为size
void heapify(int i,int size)
{
	//i为当前节点的位置 
	int l = i*2+1;
	while(l < size)
	{
		//左右孩子中比较得出最大的孩子 
		int best = l+1 < size && arr[l+1] > arr[l] ? l+1 : l;
		//左右孩子中的最大值再与父亲节点比较 
		best = arr[best] > arr[i] ? best : i;
		//左右孩子都比父亲节点小 
		if(best == i)
			break;
		swap(best,i);
		i = best;  //更新当前节点的位置 
		l = i*2+1; //更新当前节点的左孩子的位置 
	}
}


//从顶到底建立大根堆，O(n * logn)
//依次弹出堆内最大值并排好序，O(n * logn)
//整体时间复杂度O(n * logn)
void heapSort1()
{
	//建立大根堆
	for(int i = 0; i < n; i++)
		heapInsert(i);
	int size = n;
	while(size > 1)
	{
		swap(0,--size);
		heapify(0,size);
	}
} 

//从底到顶建立大根堆，O(n)
//依次弹出堆内最大值并排好序，O(n * logn)
//整体时间复杂度为O(n * logn)
void heapSort2()
{
	//完全二叉树
	//每个子树的根节点都是当前树的最大值 
	for(int i = n-1; i >= 0; i--)
		heapify(i,n);
	int size = n;   //树的节点数，弹出则减1 
	while(size > 1)
	{
		swap(0,--size);
		heapify(0,size);
	}
 } 


int main()
{
	cin >> n;
	for(int i = 0; i < n; i++)
		cin >> arr[i];
	// heapSort1();
	heapSort2();
	for(int i = 0; i < n; i++)
		cout << arr[i] << " ";
	cout << endl;
	system("pause");
	return 0;
}

File: py/排序/堆排序.py
# 堆排序
class MAXHeap:
    def __init__(self, arr):
        self.heap = arr
        self.n = len(arr)

    def _swap(self, i, j):
        self.heap[i], self.heap[j] = self.heap[j], self.heap[i]

    def _heap_insert(self, i):
        while self.heap[i] > self.heap[(i - 1) // 2]:
            self._swap(i, (i - 1) // 2)
            i = (i - 1) // 2
    
    def _heapify(self, i, size):
        l = i * 2 + 1
        while l < size:
            best = l + 1 if l + 1 < size and \
            self.heap[l + 1] > self.heap[l] else l
            best = best if self.heap[best] > self.heap[i] else i
            if best == i:
                break
            self._swap(i, best)
            i = best
            l = i * 2 + 1

    def _heap_sort(self):
        for i in range(self.n - 1, -1, -1):
            self._heapify(i, self.n)
        size = self.n
        while size > 1:
            size -= 1
            self._swap(0, size)
            self._heapify(0, size)

    def get_sorted_arr(self):
        self._heap_sort()
        return self.heap

    def get_k_smallest(self, k):
        if k > self.n:
            return "输入的k值过大"
        return self.heap[:k]


class MINHeap:
    def __init__(self, arr):
        self.heap = arr
        self.n = len(arr)

    def _swap(self, i, j):
        self.heap[i], self.heap[j] = self.heap[j], self.heap[i]
    
    def _heap_insert(self, i):
        while self.heap[i] < self.heap[(i - 1) // 2]:
            self._swap(i, (i - 1) // 2)
            i = (i - 1) // 2
        
    def _heapify(self, i, size):
        l = i * 2 + 1
        while l < size:
            best = l + 1 if l + 1 < size and \
            self.heap[l + 1] < self.heap[l] else l
            best = best if self.heap[best] < self.heap[i] else i
            if best == i:
               break
        self._swap(i, best)
        i = best
        l = i * 2 + 1

    def _heap_sort(self):
        for i in range(self.n - 1, -1, -1):
            self._heapify(i, self.n)
        size = self.n
        while size > 1:
            size -= 1
            self._swap(0, size)
            self._heapify(0, size)

    def get_sorted_arr(self):
        self._heap_sort()
        return self.heap


if __name__ == '__main__':
    arr = list(map(int, input("请输入待排序数组：\n").split()))
    heap_1 = MAXHeap(arr)
    heap_2 = MINHeap(arr)
    # print("堆排序结果(大根堆)：\n", heap_1.get_sorted_arr())
    # print("堆排序结果(小根堆)：\n", heap_2.get_sorted_arr())
    k = int(input("请输入要查找的最小的k个元素："))
    print("最小的k个元素：\n", heap_1.get_k_smallest(k))